### PR TITLE
Use coordinator to serialize data access in the SMA WA-177

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

**Comment by kshakir** [comment]
jenkins retest

File: core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -43,6 +43,19 @@ submissionmonitor {
   trackDetailedSubmissionMetrics = true
 }
 
+data-source {
+  coordinated-access {
```
Comment (salonishah11) [inline core/src/main/resources/reference.conf:47]: OOC, just like Cromwell, do we need to update rawls.conf config in firecloud-develop for this change?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/coordination/CoordinatedDataSourceActorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    val numTested = 100
+    implicit val implicitAskTimeout: Timeout = Timeout(individualWaitTimeout * numTested * 2)
+    val testActor = TestActorRef(CoordinatedDataSourceActor.props())
+    val mockSlickDataSource = mock[SlickDataSource]
+    val mockDataAccessFunction = mock[DataAccess => ReadWriteAction[Any]]
+    val transactionIsolation = TransactionIsolation.RepeatableRead
+
+    val quickening = new Quickening(individualSleep)
+    when(mockSlickDataSource.inTransaction(mockDataAccessFunction, transactionIsolation))
+      .thenAnswer(_ => Future(quickening.highlander()))
```
Comment (salonishah11) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/coordination/CoordinatedDataSourceActorSpec.scala:105]: What does `highlander()` do? I saw the code, but can't understand it's purpose

Comment (kshakir) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/coordination/CoordinatedDataSourceActorSpec.scala:105]: The function throws an exception if there was more one `.highlander()` call at the same time.

[There can be only one](https://www.youtube.com/watch?v=sqcLjcSloXs)

When run by multiple threads, the function's internal `counter` will be greater than one if the callers were not synchronized correctly.

File: core/src/main/resources/reference.conf (hunk 1 above)
Comment (kshakir) [reply core/src/main/resources/reference.conf:47]: Nope. I turned the coordination on by default on the next couple lines below.

The new knob allows us to turn off the functionality, if needed for some reason, without needing to recompile the code.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/coordination/DataSourceAccess.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                                  waitTimeout: FiniteDuration,
+                                  askTimeout: FiniteDuration,
+                                 ) extends DataSourceAccess {
+  override def inTransaction[A: ClassTag](dataAccessFunction: DataAccess => ReadWriteAction[A],
+                                          isolationLevel: TransactionIsolation): Future[A] = {
+    // How long to wait for the actor to get to and finish running the function
+    implicit val implicitTimeout: Timeout = Timeout(askTimeout)
+    // Throws a DeadlineTimeoutException the function is not started within startTimeout
+    // Throws a TimeoutException if the function does not run within the waitTimeout
+    // Throws an AskTimeoutException if we do not hear back either way before askTimeout
```
Comment (salonishah11) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/coordination/DataSourceAccess.scala:54]: Are these new exceptions? Or were they already handled in the SubmissionMonitorActor?

**Review by salonishah11 — COMMENTED** [review]
What happens when the system is under heavy load? Is it possible that many transaction requests throw `DeadlineTimeoutException` or `TimeoutException`? Would they be retried?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/coordination/DataSourceAccess.scala (hunk 3 above)
Comment (kshakir) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/coordination/DataSourceAccess.scala:54]: These are new Exceptions, however the SMA has an Akka supervisor that retries non-fatal errors.

**Comment by kshakir** [comment]
> What happens when the system is under heavy load? Is it possible that many transaction requests throw `DeadlineTimeoutException` or `TimeoutException`? Would they be retried?

The perf results are posted in the WA-177 comments. Here's a clip:

---

> In the Perf environment, 1,000 submissions were submitted using Apache Bench consisting of 3 workflows each, with the workflows instructed to call cache in cromwell.

rawls branch | max(number of submission monitor retries) | duration from start to last terminating submission actor | kibana link for start to start+90m (data expires after 30 days)
-- | -- | -- | --
dev | 44 | 01h 05m 36s | [link](https://kibana.logit.io/goto/f8b7c5c8277c2319911a65d3595d9269)
WA-177 | 32 | 01h 22m 08s | [link](https://kibana.logit.io/goto/1c824ba389acdb758d8af4b4988cc763)

**Review by salonishah11 — APPROVED** [review]
>The perf results are posted in the WA-177 comments.

@kshakir Sorry I forgot to take a look at the comments. LGTM 👍

**Review by aednichols — APPROVED** [review]
Nice!

**Pull request merged**